### PR TITLE
[docker] use shadow projects for the possibility of semi-incremental compilation.

### PR DESCRIPTION
**Pull request by rexhoffman**
## Motivation

Three valuable changes.
1: mint docker file compiles dependencies in a separate, reusable layer
2: Dockerfiles do not force recompiles on rust code.
3: mint docker-run.sh (the docker entry point) file is executed (the command is tested to start a process).

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

#start network
docker/validator-dynamic/run.sh
#start minter
docker/mint/run.sh
#start client
cargo run -p cli --bin cli -- -a localhost -p 8000 -f localhost:8080
#create account and add coins
a c #account create
a list #account list
a mint <account number> 100 #mint 100 coins in account.... 
q b <account number> #see account balance.
## Related PRs

none

### CODE REVIEW

**Comment by sausagee** [comment]
Chatted with Rex offline about the changes. I feel like we need to split this PR in two parts to land the proposed changes smoothly: one for the test integration and another for the build optimization. 

On the build optimization part. I feel like we need two additional changes: 
- apply the same change to other Docker builds to avoid layer cache miss 
- find a place to host the shadow deps instead of Jack's repo (this is to avoid single point of failure in build infra)

**Comment by sausagee** [comment]
cc @mgorven to take another on the build and pre-test layer change.

File: docker/mint/mint.Dockerfile [hunk 1]
```diff
@@ -31,11 +70,41 @@ RUN pip3 install -r /libra/docker/mint/requirements.txt
...
+# GIVEN
+ARG IMAGE="libra_mint:latest"
+ARG AC_HOST="172.18.0.13"
+ARG AC_PORT="8000"
+ARG LOG_LEVEL="info"
+ARG MINT_KEY="1337133713371337133713371337133713371337"
+# WHEN
+RUN /opt/libra/bin/docker-run.sh & \
+   while [ ! -f /tmp/pidfile ]; do sleep 1; done; \
+   PIDS=$( cat /tmp/pidfile ); \
```
Comment (sausagee) [inline docker/mint/mint.Dockerfile]: I'm wonder if there is another way to find the PID of gunicorn without writing it to file in `docker-run.sh`.

File: docker/mint/mint.Dockerfile [hunk 2]
```diff
@@ -31,11 +70,41 @@ RUN pip3 install -r /libra/docker/mint/requirements.txt
...
+# time out.
+
+# Necessary environment variabls for gunicorn (in docker-run.sh)
+# Setup and perform the test
+# GIVEN
+ARG IMAGE="libra_mint:latest"
+ARG AC_HOST="172.18.0.13"
+ARG AC_PORT="8000"
+ARG LOG_LEVEL="info"
+ARG MINT_KEY="1337133713371337133713371337133713371337"
```
Comment (sausagee) [inline docker/mint/mint.Dockerfile]: IIRC docker-run.sh uses `$CFG_SEED` instead of `$MINT_KEY`?

File: .dockerignore [hunk 3]
```diff
@@ -3,3 +3,4 @@
 **/.terraform/
 target/
 !target/libra-node-builder/libra-node
+**/*Dockerfile
```
Comment (mgorven) [inline .dockerignore]: This breaks layer caching between the different images. The Dockerfile getting used is added to the context, which changes the context hash for the different image builds.

**Comment by mgorven** [comment]
Wow this is complex. I need to find time to digest this properly.

File: .dockerignore (hunk 3 above)
Comment (rexhoffman) [reply .dockerignore]: so including all the docker files as source (for the copy) is a means to share layers between docker builds with different different docker files, but with identical lines?  Got it.

What other ways of achieving this exist?   Perhaps an image with all compiled code in it used as a FROM?   Looking for ways to share layers w/o having a change to a docker file force cargo to recompile.

Also contemplating building nightlies images that retain the cargo target dirs for reuse (much the same way local build works, or the shadow image's target is reused in this pr).

File: docker/mint/mint.Dockerfile (hunk 1 above)
Comment (rexhoffman) [reply docker/mint/mint.Dockerfile]: My typical ps -ef | grep tool wasn't on the path, and was reluctant to add it - but can for the test.
Not sure why we need gunicorn today.   Meaning to read through our python server file at some point soon.

File: docker/mint/mint.Dockerfile (hunk 2 above)
Comment (rexhoffman) [reply docker/mint/mint.Dockerfile]: I was simply reusing what is in the run.sh.   One of the things I intend to clean up, but didn't want to flip too many stones at once.

File: docker/mint/mint.Dockerfile [hunk 4]
```diff
@@ -8,18 +8,57 @@ RUN apt-get update && apt-get install -y cmake curl clang git
...
+ENV DEP_SHADOW_VERSION 956632e3c7b2f501758de6a92f38c50cd3fffaf3
+RUN cargo install --git https://github.com/metajack/cargo-create-deponly-shadow --rev ${DEP_SHADOW_VERSION}
 
-RUN cargo build --release -p libra-node -p cli -p config-builder -p safety-rules && cd target/release && rm -r build deps incremental
+# Here we start trickery
+# Copy in the code.
+WORKDIR /libra
+COPY . /libra
+# Generate a shadow project (Cargo.toml(s)/Cargo.lock/needed rust files.)
+# not that the rust files are currently generated with current timestamp,
```
Comment (rexhoffman) [inline docker/mint/mint.Dockerfile]: typo.

File: docker/mint/mint.Dockerfile (hunk 1 above)
Comment (sausagee) [reply docker/mint/mint.Dockerfile]: hmm... I think it's worth to add those basic utils. Just not a fan of writing PID to file.

File: .dockerignore (hunk 3 above)
Comment (sausagee) [reply .dockerignore]: IIUC, this should avoid rebuilding the expensive rust build step, esp when iterating changes on Docker files.

**Pull request closed without merging**